### PR TITLE
Update food checkout confirmation and order claim pages

### LINKED ISSUE #621: Support offerings with no fulfillments
- If an offering has one fulfillment, initialize a checkout using it (like we do if there is a previous checkout)
- If an offering has no available fulfillments:
   - hide the section entirely in the UI
   - do not require the parameter (in fact it cannot be sent- if it were sent, we'd 400 for an invalid fulfillment id)

### LINKED ISSUE #626: Update food checkout flow pages
The confirmation page is confusing to some members, Malena and Jacqueline drafted a mock-page of how the Food Confirmation and Order Claim page can look like to facilitate the pickup process and not confuse members too much.

- [ ] Mock page as closely as possible to the 'Food confirmation page draft'
  - [ ] Check if we can setup dynamic content rather than hard-coded content
- [ ] 'order claim page', move the press and hold button to the top, show rest of details at the bottom

### Food confirmation page draft
Screenshot 2024-04-18 at 2 47 37 PM

### Order claim page draft
Screenshot 2024-04-16 at 12 15 05 PM

### PR DESCRIPTION
**Pull request by DeeTheDev**
Fixes #626
Fixes #621 

Update the food checkout confirmation and order claim pages in accordance to the issue's image drafts.
The offering requires an additional field `fulfillment_instructions` which is used in the confirmation page to show more detailed instructions for the item pickup, e.g. default pickup location etc. It also appears below the fulfillment prompt section in the food checkout page (if it is available).
- Move order 'Press and Hold' button to the top so that members can access it quicker on page load

---

Adds support for offerings with no fulfillment options. Since members reportedly get confused with the fulfillment prompt/options, it makes sense to make these fields (and the checkout fulfillment section) optional.

---

Support the above changes in the adminapp. Including updating the offering route parameter validations correctly and updating the create/update/read routes.
- Update OfferingForm layout structure to make sense of the above changes/better eye-flow

Screenshot 2024-04-23 at 6 15 20 PM
Screenshot 2024-04-23 at 6 15 35 PM

### Food checkout fulfillment section update
First fulfillment is used if it is the only one available in the offering. Fulfillment instructions below prompt.
Screenshot 2024-04-23 at 6 17 03 PM

### Food checkout confirmation update
Screenshot 2024-04-23 at 6 17 28 PM
Screenshot 2024-04-23 at 6 17 36 PM

### Claim order 'press and hold' button moved to top
Screenshot 2024-04-23 at 6 28 57 PM

### CODE REVIEW

File: db/migrations/045_fulfillment_instructions.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+Sequel.migration do
+  change do
+    alter_table(:commerce_offerings) do
+      add_foreign_key :fulfillment_instructions_id, :translated_texts, null: false
```
Comment (rgalanakis) [inline db/migrations/045_fulfillment_instructions.rb:6]: What's going to happen when you run this on production with `null: false`?

File: lib/suma/admin_api/commerce_offerings.rb [hunk 2]
```diff
@@ -80,14 +81,24 @@ class PicklistEntity < BaseEntity
...
+      DetailedOfferingEntity,
     ) do
       params do
         requires :image, type: File
         requires(:description, type: JSON) { use :translated_text }
-        requires(:fulfillment_prompt, type: JSON) { use :translated_text }
-        requires(:fulfillment_confirmation, type: JSON) { use :translated_text }
-        requires :fulfillment_options,
+        requires :fulfillment_prompt, type: JSON do
+          requires :en, type: String, allow_blank: true
```
Comment (rgalanakis) [inline lib/suma/admin_api/commerce_offerings.rb:90]: Why are you copy-pasting this rather than `use :translated_text` as it was?

File: lib/suma/admin_api/commerce_offerings.rb [hunk 3]
```diff
@@ -102,18 +113,28 @@ class PicklistEntity < BaseEntity
...
       Suma::Commerce::Offering,
-      DetailedEntity,
+      DetailedOfferingEntity,
     ) do
       params do
         optional :image, type: File
         optional(:description, type: JSON) { use :translated_text }
-        optional(:fulfillment_prompt, type: JSON) { use :translated_text }
-        optional(:fulfillment_confirmation, type: JSON) { use :translated_text }
+        optional :fulfillment_prompt, type: JSON do
```
Comment (rgalanakis) [inline lib/suma/admin_api/commerce_offerings.rb:126]: Same here, why the change?

File: lib/suma/api/commerce.rb [hunk 4]
```diff
@@ -95,8 +95,10 @@ def lookup_cart!(offering)
               payment_instrument: member.default_payment_instrument,
               save_payment_instrument: member.default_payment_instrument.present?,
             )
-            if existing_editable_checkout&.fulfillment_option
-              checkout.update(fulfillment_option: existing_editable_checkout.fulfillment_option)
+            if offering.fulfillment_options.one?
+              checkout.update(fulfillment_option: offering.fulfillment_options.first)
+            elsif (fulfillment_option = existing_editable_checkout&.fulfillment_option)
+              checkout.update(fulfillment_option:)
```
Comment (rgalanakis) [inline lib/suma/api/commerce.rb:101]: Needs test coverage.

File: lib/suma/api/commerce.rb [hunk 5]
```diff
@@ -165,7 +167,7 @@ def lookup_editable!
             checkout.payment_instrument = instrument if instrument
           end
 
-          if params.key?(:fulfillment_option_id)
+          if params[:fulfillment_option_id]
```
Comment (rgalanakis) [inline lib/suma/api/commerce.rb:170]: Why this change?

File: webapp/src/pages/FoodCheckout.jsx [hunk 6]
```diff
@@ -361,7 +362,7 @@ function OrderSummary({ checkout, chosenInstrument }) {
   // and if there's an error we'll deal with it.
   const showSubmit = checkout.checkoutProhibitedReason !== "charging_prohibited";
   return (
-    <Col xs={12}>
+    <LayoutContainer gutters className="my-4">
```
Comment (rgalanakis) [inline webapp/src/pages/FoodCheckout.jsx:365]: Why are all of these switched to be layout containers? Usually you only want something like that at the root; this is going to be a pain to rearrange, I think.

File: webapp/src/pages/FoodCheckoutConfirmation.jsx [hunk 7]
```diff
@@ -49,30 +51,56 @@ export default function FoodCheckoutConfirmation() {
...
+          <LayoutContainer gutters>
+            <h4>{offering.fulfillmentConfirmation}</h4>
+            <p>{fulfillmentOption.description}</p>
+          </LayoutContainer>
+          <hr className="my-4" />
+        </>
+      )}
       <LayoutContainer gutters>
-        <p>{t("food:confirmation_message")}</p>
+        <h4>{t("food:confirmation_transportation_title")}</h4>
```
Comment (rgalanakis) [inline webapp/src/pages/FoodCheckoutConfirmation.jsx:85]: This is not suitable to hard-code into the UI. It needs to be done using the confirmation markdown.

File: webapp/src/pages/Mobility.jsx [hunk 8]
```diff
@@ -12,6 +13,17 @@ export default function Mobility() {
       <LayoutContainer top gutters>
         <h5>{t("mobility:title")}</h5>
         <p className="text-secondary">{t("mobility:intro")}</p>
+        <hr />
+        {mdx("mobility:ride_connection_notice", {
```
Comment (rgalanakis) [inline webapp/src/pages/Mobility.jsx:17]: This should be put into `mobility:intro`. There is a difference between putting stuff like this in localization data (which is just a file that can be swapped out) and in the actual code (which cannot vary or be modified externally).

**Review by rgalanakis — CHANGES_REQUESTED** [review]

File: webapp/src/pages/FoodCheckout.jsx (hunk 6 above)
Comment (DeeTheDev) [reply webapp/src/pages/FoodCheckout.jsx:365]: I contemplated with this too, I'll revert this layout change

File: lib/suma/admin_api/commerce_offerings.rb (hunk 2 above)
Comment (DeeTheDev) [reply lib/suma/admin_api/commerce_offerings.rb:90]: Since we allow to pass `fulfillment_prompt: {en: "", es: ""}` it will, translated_text doesn't allow_blank strings, I updated translated_text to pass this option through args

File: webapp/src/pages/Mobility.jsx [hunk 9]
```diff
@@ -1,17 +1,27 @@
...
 export default function Mobility() {
   return config.featureMobility ? (
     <>
       <LayoutContainer top gutters>
         <h5>{t("mobility:title")}</h5>
-        <p className="text-secondary">{t("mobility:intro")}</p>
+        {mdx("mobility:intro", {
+          overrides: {
+            a: { component: MdLink },
+            p: {
```
Comment (rgalanakis) [inline webapp/src/pages/Mobility.jsx:18]: Nicely done.

File: webapp/src/pages/Mobility.jsx [hunk 10]
```diff
@@ -1,17 +1,27 @@
...
 
 export default function Mobility() {
   return config.featureMobility ? (
     <>
       <LayoutContainer top gutters>
         <h5>{t("mobility:title")}</h5>
-        <p className="text-secondary">{t("mobility:intro")}</p>
+        {mdx("mobility:intro", {
+          overrides: {
+            a: { component: MdLink },
```
Comment (rgalanakis) [inline webapp/src/pages/Mobility.jsx:17]: We may want to put this into `mdx` itself. To consider for the future.

File: webapp/src/pages/FoodCheckoutConfirmation.jsx [hunk 11]
```diff
@@ -49,53 +50,77 @@ export default function FoodCheckoutConfirmation() {
         <p className="mb-0">{t("food:confirmation_subtitle")}</p>
       </div>
       <LayoutContainer gutters top>
-        <h4 className="mb-3">{t("food:checkout_items_title")}</h4>
+        <h4 className="mb-3">{t("food:confirmation_my_order")}</h4>
         {items.map((p, idx) => (
           <Item key={idx} item={p} />
         ))}
+        {user.unclaimedOrdersCount !== 0 && (
+          <FormButtons
```
Comment (rgalanakis) [inline webapp/src/pages/FoodCheckoutConfirmation.jsx:58]: Make this a regular button, using flexbox to lay it out as you normally would. This is not a form.

If you start to hijack something semantically distinct (like form buttons) for something different (like a link button as you have here), it makes refactoring and redesigning a much bigger headache.

File: webapp/src/pages/FoodCheckoutConfirmation.jsx [hunk 12]
```diff
@@ -49,53 +50,77 @@ export default function FoodCheckoutConfirmation() {
...
+            <p>{fulfillmentOption.description}</p>
+          </LayoutContainer>
+          <hr className="my-4" />
+        </>
+      )}
       <LayoutContainer gutters>
-        <p>{t("food:confirmation_message")}</p>
+        <h4>{t("food:confirmation_transportation_title")}</h4>
+        <p className="mb-0">{t("food:confirmation_transportation_subtitle")}</p>
         <FormButtons
```
Comment (rgalanakis) [inline webapp/src/pages/FoodCheckoutConfirmation.jsx:86]: As above, this is not really a form either. There's no form, inputs, or secondary action that would be a reason to use this component. It's just a single centered button, AFAICT.

File: webapp/src/pages/FoodCheckout.jsx [hunk 13]
```diff
@@ -148,26 +147,30 @@ export default function FoodCheckout() {
...
-                ...attrs,
-              })
-            }
-          />
-        </LayoutContainer>
-        <hr />
-        <LayoutContainer gutters className="mb-4 mt-4">
+        {!isEmpty(checkout.availableFulfillmentOptions) && (
+          <>
+            <LayoutContainer gutters className="my-4">
```
Comment (rgalanakis) [inline webapp/src/pages/FoodCheckout.jsx:152]: I don't understand why we are using nested containers like this. It's a bad practice, as per the bootstrap docs: https://getbootstrap.com/docs/5.0/layout/containers/

If something is called 'layout' you normally only have one of them.

This is too big of an issue to fix now but please do not continue using LayoutContainer (or nested Containers) everywhere. If you want flexbox, just use flexbox. If you want a grid, use bootstrap row/col. I think the changes here, removing all the grid components, are largely a step backwards, but 🤷

File: webapp/src/components/OrderDetail.jsx [hunk 14]
```diff
@@ -105,6 +105,10 @@ function FulfillmentOption({ order, onOrderUpdated }) {
   const [optionId, setOptionId] = React.useState(0);
   const { showErrorToast } = useErrorToast();
 
+  if (!order.fulfillmentOption) {
```
Comment (rgalanakis) [inline webapp/src/components/OrderDetail.jsx:108]: Someone still needs to be able to edit if there is no chosen fulfillment option, but there are options to choose. Maybe you want this inside the `isEmpty`check on line 112?

File: lib/suma/admin_api/commerce_offerings.rb [hunk 15]
```diff
@@ -80,14 +81,15 @@ class PicklistEntity < BaseEntity
...
-      DetailedEntity,
+      DetailedOfferingEntity,
     ) do
       params do
         requires :image, type: File
         requires(:description, type: JSON) { use :translated_text }
-        requires(:fulfillment_prompt, type: JSON) { use :translated_text }
-        requires(:fulfillment_confirmation, type: JSON) { use :translated_text }
-        requires :fulfillment_options,
+        optional(:fulfillment_prompt, type: JSON) { use :translated_text, allow_blank: true  }
```
Comment (rgalanakis) [inline lib/suma/admin_api/commerce_offerings.rb:89]: Why do we only want allow_blank: true here and not everywhere these params are used?

File: spec/suma/api/commerce_spec.rb [hunk 16]
```diff
@@ -391,13 +378,6 @@
       expect(checkout.refresh).to have_attributes(fulfillment_option: be === opt)
     end
 
-    it "errors if a nil fulfillment empty is passed" do
```
Comment (rgalanakis) [inline spec/suma/api/commerce_spec.rb:394]: Why was this removed?

File: spec/suma/commerce/cart_spec.rb [hunk 17]
```diff
@@ -219,4 +219,59 @@ def create_fake_order(cart)
       expect(described_class[item.id]).to have_attributes(offering_product: be === op) # non-eager loading
     end
   end
+
+  describe "create_checkout" do
```
Comment (rgalanakis) [inline spec/suma/commerce/cart_spec.rb:223]: Port over all the specs, not just a handful of them. Otherwise you'll have to do code archaeology to see what was done, when you have to maintain or edit this or the API code.

File: lib/suma/admin_api/commerce_offerings.rb (hunk 15 above)
Comment (DeeTheDev) [reply lib/suma/admin_api/commerce_offerings.rb:89]: @rgalanakis In some cases, I think there shouldn't be empty :translated_text such as payment triggers `receiving_ledger_contribution_text` param. Ultimately we could set all to allow_blank but I'm not sure if thats a good idea.

File: webapp/src/pages/Mobility.jsx (hunk 10 above)
Comment (DeeTheDev) [reply webapp/src/pages/Mobility.jsx:17]: I think that's a good idea. will add it to mdx when i change the mobility content again.

File: webapp/src/components/OrderDetail.jsx (hunk 14 above)
Comment (DeeTheDev) [reply webapp/src/components/OrderDetail.jsx]: This made sense, and works like we expect now. Thanks.

File: spec/suma/commerce/cart_spec.rb (hunk 17 above)
Comment (DeeTheDev) [reply spec/suma/commerce/cart_spec.rb:223]: I think I ported all specs now, the API specs left-over are specific to the commerce API. I did update the specs though with commit [fc68742](https://github.com/lithictech/suma/pull/629/commits/fc68742a59bc441744a27ac92f5ef8649ddf6618)

File: spec/suma/api/commerce_spec.rb (hunk 16 above)
Comment (DeeTheDev) [reply spec/suma/api/commerce_spec.rb:394]: Error on my part. I updated the test with "..allows nil fulfillment option".

File: lib/suma/commerce/order.rb [hunk 18]
```diff
@@ -216,7 +219,9 @@ def can_begin_fulfilling_on_create?(now: Time.now)
   end
 
   def can_claim?
-    return self.fulfillment_status == "fulfilling" && self.checkout.fulfillment_option&.pickup?
+    checkout_option = self.checkout.fulfillment_option
+    return self.fulfillment_status == "fulfilling" && checkout_option.pickup? unless checkout_option.nil?
```
Comment (rgalanakis) [inline lib/suma/commerce/order.rb:223]:
```rb
return false unless self.fulfillment_status == 'fulfilling'
checkout_option = self.checkout.fulfillment_option
# There isn't a logical reason 'no fulfillment option' means 'claimable',
# but that's what we're going with for now.
return checkout_option.nil? || checkout_option.pickup?
```

File: spec/suma/commerce/order_spec.rb [hunk 19]
```diff
@@ -140,7 +140,7 @@
 
       order.refresh
       order.checkout.fulfillment_option = nil
-      expect(order).to_not be_can_claim
+      expect(order).to be_can_claim
```
Comment (rgalanakis) [inline spec/suma/commerce/order_spec.rb:143]: The spec description is now wrong. You can split this into two specs.
